lib/poison/decoder.ex

```diff
@@ -40,9 +40,13 @@ defmodule Poison.Decode do
   end
 
   defp transform_struct(value, _keys, as, options) do
-    kv = for k <- Map.keys(Map.from_struct(as.__struct__)) do
-      {k, Map.get(value, Atom.to_string(k))}
-    end
+    kv = Map.keys(Map.from_struct(as.__struct__))
+          |> Enum.reduce([], fn key, acc ->
+               case Map.get(value, Atom.to_string(key)) do
+                 nil -> acc
+                 value -> [{key, value} | acc]
+               end
+             end)
```

Comment:
> Is there a way to make this better?

Remove the `:__struct__` key and then re-add it after running `Enum.into` it also means we don't need the first clause.

```elixir
struct = as.__struct__.__struct__
Enum.into(Map.from_struct(as.__struct__), %{}, fn {key, struct_value} ->
  case Map.fetch(value, Atom.to_string(key)) do
    {:ok, json_value} -> {key, json_value}
    :error -> {key, struct_value}
  end
end)
|> Map.put(:__struct__, struct)
|> Poison.Decoder.decode(options)
```

Something like this maybe.